micrometer-observation/src/main/java/io/micrometer/observation/Observation.java

```diff
@@ -587,13 +561,44 @@ default <T> Supplier<T> wrap(Supplier<T> supplier) {
      * @param <E> type of exception checkedCallable throws
      * @return the result from {@link CheckedCallable#call()}
      */
-    @SuppressWarnings("unused")
+    @Nullable
     default <T, E extends Throwable> T observeChecked(CheckedCallable<T, E> checkedCallable) throws E {
+        return observeCheckedWithContext((context) -> checkedCallable.call());
+    }
+
+    default <T, E extends Throwable> CheckedCallable<T, E> wrapChecked(CheckedCallable<T, E> checkedCallable) throws E {
+        return () -> observeChecked(checkedCallable);
+    }
+
+    /**
+     * Observes the passed {@link Function} which provides access to the {@link Context}.
+     *
+     * This means the followings:
+     * <ul>
+     * <li>Starts the {@code Observation}</li>
+     * <li>Opens a {@code Scope}</li>
+     * <li>Calls {@link Function#apply(Object)} where it gets a {@link Context}</li>
+     * <li>Closes the {@code Scope}</li>
+     * <li>Signals the error to the {@code Observation} if any</li>
+     * <li>Stops the {@code Observation}</li>
+     * </ul>
+     *
+     * NOTE: When the {@link ObservationRegistry} is a noop, this function receives a
+     * default {@link Context} instance which is not the one that have passed at
+     * {@link Observation} creation.
+     * @param function the {@link Function} to call
+     * @return the result from {@link Function#apply(Object)}
+     * @param <C> the type of input {@link Context} to the function
+     * @param <T> the type parameter of the {@link Function} return
+     */
+    @SuppressWarnings({ "unchecked" })
+    @Nullable
+    default <C extends Context, T> T observeWithContext(Function<C, T> function) {
         start();
         try (Scope scope = openScope()) {
-            return checkedCallable.call();
+            return function.apply((C) getContext());
         }
-        catch (Throwable error) {
+        catch (Exception error) {
```

Comment:
Since its pair (`observeCheckedWithContext`) catches `Throwable`, should not we catch that here too?